packages/office-ui-fabric-react/src/components/Pivot/Pivot.base.tsx

```diff
@@ -41,53 +39,32 @@ export class PivotBase extends BaseComponent<IPivotProps, IPivotState> {
   private _pivotId: string;
   private focusZone = React.createRef<FocusZone>();
   private _classNames: { [key in keyof IPivotStyles]: string };
+  private _links: IPivotItemProps[] = [];
+  private _previousProps: IPivotProps | undefined = undefined;
```

Comment:
I realize the code that was originally here was perhaps not the best implementation. For example, does this component really care about any characteristic of props changing, or is this just a brute force method?

In any case, I'm not sure if adding a props class member is the right approach for replacing `componentWillReceiveProps`. I'm worried it's basically simulating what `componentWillReceiveProps` does, potentially introducing the same issues with async rendering. I don't think this is a pattern we'd want others replicating. Looking at the [documentation](https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops) it seems like a memoization helper is another option. Is that an option for us here? What about [getDerivedStateFromProps](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops)?

Reply:
From a conversation we had in person, I think @jdhuntington's concern about getDerivedStateFromProps is that it may be re-run every render (not just when props change), making it more expensive than it seems on the surface. But I don't think the current approach is the right way to work around that concern.

Reply:
Though one thing this PR is right about is that `links` and `selectedTabId` don't belong in state--per [this article](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#anti-pattern-unconditionally-copying-props-to-state), state is not meant as a place for cached calculation results.

This component also has the problem (as do many of our components...) that it's a weird hybrid of controlled and uncontrolled, which makes a lot of logic messy (see "Anti-pattern: erasing state when props change" in the linked article). I'm pretty sure that this change in its current state will make the pivot *only* work as controlled, since it unconditionally prefers the selectedKey value from props rather than state.

Reply:
Here's a demo of when getDerivedStateFromProps is called. Seems it is in fact before every render regardless of whether props changed. 
https://codepen.io/ecraig12345/pen/RvQxmZ

Reply:
Yeah, it seemed on the surface that a memoization helper was the more applicable solution here but wasn't 100% positive. I think it'd be good to have a memoization helper implementation for reference since afaik we don't really have many examples currently.

Reply:
While we're on the topic, there are two catches with memoized functions: 

1. Per our memoize decorator's documentation, you can't access `this` within a memoized function. (For that reason, I've been leaning towards putting memoized functions outside the classes and using `memoizeFunction()` rather than `@memoize`--because there's NO compile time warning that `this` won't be accessible from a memoized private method.)

2. You can't pass objects (like `props`) as params, you have to pass each value the function cares about as a separate param. 
```
const badMemoized = memoizeFunction((props: IFooProps) => props.foo + props.bar);

const goodMemoized = memoizeFunction((foo: number, bar: number) => foo + bar);
```